src/backend/EdFi.DataManagementService.Backend.Postgresql/Operation/UpsertDocument.cs

```diff
@@ -93,6 +107,19 @@ NpgsqlTransaction transaction
                 return new UpsertResult.UpsertFailureWriteConflict();
             }
 
+            if (pe.SqlState == PostgresErrorCodes.UniqueViolation)
+            {
+                _logger.LogInformation(
+                    "Failure: alias identity already exists - {TraceId}",
+                    upsertRequest.TraceId
+                );
+                var duplicates =
+                    upsertRequest.DocumentInfo.DocumentIdentity.DocumentIdentityElements.Select(d =>
+                        $"({d.IdentityJsonPath.Value.Substring(d.IdentityJsonPath.Value.LastIndexOf('.') + 1)} = {d.IdentityValue})");
+                return new UpsertResult.UpsertFailureIdentityConflict(
+                    $"A natural key conflict occurred when attempting to create a new resource {upsertRequest.ResourceInfo.ResourceName.Value} with a duplicate key. The duplicate keys and values are {string.Join(',', duplicates)}");
```

Comment:
I think we'll want the message constructed in DMS core rather than each backend, so how about changing `UpsertResult.UpsertFailureIdentityConflict` to also take a `ResourceName` and change `ReferencingDocumentInfo` to be your `duplicates` array.

Reply:
Good call. I went with a generic `IEnumerable<KeyValuePair<string, string>>` to allow some flexibility in the error message.